drivers/sensor/iis2mdc/iis2mdc.c

```diff
@@ -367,6 +336,119 @@ static int iis2mdc_init(const struct device *dev)
 	return 0;
 }
 
-DEVICE_DT_INST_DEFINE(0, iis2mdc_init, device_pm_control_nop,
-		     &iis2mdc_data, &iis2mdc_dev_config, POST_KERNEL,
-		     CONFIG_SENSOR_INIT_PRIORITY, &iis2mdc_driver_api);
+#if DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 0
+#warning "IIS2MDC driver enabled without any devices"
+#endif
+
+/*
+ * Device creation macro, shared by IIS2MDC_DEFINE_SPI() and
+ * IIS2MDC_DEFINE_I2C().
+ */
+
+#define IIS2MDC_DEVICE_INIT(inst)					\
+	DEVICE_DT_INST_DEFINE(inst,					\
+			    iis2mdc_init,				\
+			    device_pm_control_nop,			\
+			    &iis2mdc_data_##inst,			\
+			    &iis2mdc_config_##inst,			\
+			    POST_KERNEL,				\
+			    CONFIG_SENSOR_INIT_PRIORITY,		\
+			    &iis2mdc_driver_api);
+
+/*
+ * Instantiation macros used when a device is on a SPI bus.
+ */
+
+#define IIS2MDC_HAS_CS(inst) DT_INST_SPI_DEV_HAS_CS_GPIOS(inst)
+
+#define IIS2MDC_DATA_SPI_CS(inst)					\
+	{ .cs_ctrl = {							\
+		.gpio_pin = DT_INST_SPI_DEV_CS_GPIOS_PIN(inst),		\
+		.gpio_dt_flags = DT_INST_SPI_DEV_CS_GPIOS_FLAGS(inst),	\
+		},							\
+	}
+
+#define IIS2MDC_DATA_SPI(inst)						\
+	COND_CODE_1(IIS2MDC_HAS_CS(inst),				\
+		    (IIS2MDC_DATA_SPI_CS(inst)),			\
+		    ({}))
+
+#define IIS2MDC_SPI_CS_PTR(inst)					\
+	COND_CODE_1(IIS2MDC_HAS_CS(inst),				\
+		    (&(iis2mdc_data_##inst.cs_ctrl)),			\
+		    (NULL))
+
+#define IIS2MDC_SPI_CS_LABEL(inst)					\
+	COND_CODE_1(IIS2MDC_HAS_CS(inst),				\
+		    (DT_INST_SPI_DEV_CS_GPIOS_LABEL(inst)), (NULL))
+
+#define IIS2MDC_SPI_CFG(inst)						\
+	(&(struct iis2mdc_spi_cfg) {					\
+		.spi_conf = {						\
+			.frequency =					\
+				DT_INST_PROP(inst, spi_max_frequency),	\
+			.operation = (SPI_WORD_SET(8) |			\
+				      SPI_OP_MODE_MASTER |		\
+				      SPI_LINES_SINGLE |		\
+				      SPI_MODE_CPOL |			\
+				      SPI_MODE_CPHA),			\
+			.slave = DT_INST_REG_ADDR(inst),		\
+			.cs = IIS2MDC_SPI_CS_PTR(inst),		\
```

Comment:
I'd personally choose something like:
```c
		.spi_cs_label = UTIL_AND(                       \
			DT_INST_SPI_DEV_HAS_CS_GPIOS(n),        \
			DT_INST_SPI_DEV_CS_GPIOS_LABEL(n)),     \
		.spi_cs_pin = UTIL_AND(                         \
			DT_INST_SPI_DEV_HAS_CS_GPIOS(n),        \
			DT_INST_SPI_DEV_CS_GPIOS_PIN(n)),       \
		.spi_cs_flags = UTIL_AND(                       \
			DT_INST_SPI_DEV_HAS_CS_GPIOS(n),        \
			DT_INST_SPI_DEV_CS_GPIOS_FLAGS(n)),     \
```

in order to reduce the number of macros. non-blocking, though

Reply:
I though @mbolivar-nordic had something in works or merged to deal with this?  Yes, it's some commits in #30536.

Maybe we should get that merged first; it's got a lot of stuff not directly related to BME280.

Reply:
@pabigot you're right. If #30536 is going to be merged soon I'd wait for it and rework this PR using the new helpers.

Reply:
#30536 looks interesting. I see some macros reusable in this context (SPI_CS_CONTROL_XXX).

Reply:
We can wait for it to be merged first.

Reply:
It's merged now, feel free to update this PR.

Reply:
> It's merged now, feel free to update this PR.

Sure, I'll work on it asap.